pinot-core/src/main/java/org/apache/pinot/core/operator/OrderByUtils.java

```diff
@@ -0,0 +1,240 @@
+    for (OrderByInfo orderByInfo : orderByInfos) {
+      OrderType orderType = orderByInfo.getOrderType();
+
+      Comparator<GroupByRecord> comparator;
+      if (orderType.equals(OrderType.AGGREGATION_VALUE)) {
+        comparator = getAggregationComparator(orderByInfo, isIntermediate);
+      } else {
+        comparator = getGroupByKeyComparator(orderByInfo);
+      }
+      if (globalComparator == null) {
+        globalComparator = comparator;
+      } else {
+        globalComparator = globalComparator.thenComparing(comparator);
+      }
+    }
+    return globalComparator;
+  }
+
+  /**
+   * Constructs the comparator for ordering by an aggregation result
+   *
+   * When ordering intermediate results, check if the intermediate results are comparable.
+   * If they are not, extract final result for doing the comparison
+   */
+  private static Comparator<GroupByRecord> getAggregationComparator(OrderByInfo orderByInfo, boolean isIntermediate) {
+    int index = orderByInfo.getIndex();
+    AggregationFunction aggregationFunction = orderByInfo.getAggregationFunction();
+    boolean ascending = orderByInfo.isAscending();
+
+    Comparator<GroupByRecord> comparator;
+
+    // intermediate aggregation results may or may not be comparable
+    // final aggregation results are always comparable
+    if (isIntermediate && !aggregationFunction.isIntermediateResultComparable()) {
+      if (ascending) {
+        comparator = (v1, v2) -> ComparableComparator.getInstance()
+            .compare(aggregationFunction.extractFinalResult(v1.getAggregationResults()[index]),
+                aggregationFunction.extractFinalResult(v2.getAggregationResults()[index]));
+      } else {
+        comparator = (v1, v2) -> ComparableComparator.getInstance()
+            .compare(aggregationFunction.extractFinalResult(v2.getAggregationResults()[index]),
+                aggregationFunction.extractFinalResult(v1.getAggregationResults()[index]));
+      }
+    } else {
+      if (ascending) {
+        comparator = (v1, v2) -> ComparableComparator.getInstance()
+            .compare(v1.getAggregationResults()[index], v2.getAggregationResults()[index]);
+      } else {
+        comparator = (v1, v2) -> ComparableComparator.getInstance()
+            .compare(v2.getAggregationResults()[index], v1.getAggregationResults()[index]);
+      }
+    }
+    return comparator;
+  }
+
+  /**
+   * Constructs the comparator for ordering by a group by key
+   */
+  private static Comparator<GroupByRecord> getGroupByKeyComparator(OrderByInfo orderByInfo) {
+    int index = orderByInfo.getIndex();
+    ColumnDataType columnDataType = orderByInfo.getColumnDataType();
+    boolean ascending = orderByInfo.isAscending();
+
+    Comparator<GroupByRecord> comparator;
+
+    switch (columnDataType) {
+      case INT:
+        if (ascending) {
+          comparator = Comparator.comparing(GroupByRecord::getGroupByKey,
+              Comparator.comparingInt(v -> Integer.valueOf(v[index])));
```

Comment:
Yes, that is correct. However, that is needed to ensure we get the right ordering, for example, 1000 > 200 if the data type is int, and "1000" < "200" if data type is String.
The cleaner way to address this would be to not convert the group by keys into String at all (in the group-key generator), but that is a much bigger change.
If you have any suggestion on solving the right ordering without the conversion penalty in the interim, that would be great.